File: src/com/javascript-practice/binarytree/binarySearch.js
/**
 * Write a function, binarySearch, that takes in a sorted array of numnbers and target.
 * The function should return the index where the target can be found within the array.
 * If the target is not found in  the array, then return -1
 * n = length of numbers array
    Time: O(logn)
    Space: O(1)

    binarySearch([0, 1, 2, 3, 4, 5, 6, 7, 8], 6); // -> 6
 */

const binarySearch = (numbers, target) => {

    return search(numbers, 0, numbers.length - 1, target)
    
}

const search = (numbers, low, high, target) => {

    if (low > high) {
        return -1;
    }

        let mid = low + Math.floor((high - low) / 2);
        if (target < numbers[mid]) {
            return search(numbers, low, mid - 1, target);;
        } else if (target > numbers[mid]) {
            return search(numbers, mid + 1, high, target)
        }
        else {
            return mid;
        }
    
    return -1;
}

 const position = binarySearch([0, 6, 8, 12, 16, 19, 20, 28], 12); // -> 2
// const position = binarySearch([0, 6, 8, 12, 16, 19, 20, 24, 28], 27); // -> -1

console.log(position);

File: src/com/gfg/Searching/BinarySearch.java
package com.gfg.Searching;

public class BinarySearch {
    public int bSearch(int[] arr, int target) {
        int low = 0, high = arr.length;
        while (low < high) {
            int mid = low + (high - low)/2;
            if(target > arr[mid]) {
                low = mid + 1;
            }else if(target < arr[mid]) {
                high = mid - 1;
            } else {
                return mid;
            }
        }
        return  -1;
    }
    public int bSearchRecursive(int[] arr, int target) {

       return helper(arr, target,0, arr.length);
    }
    public int helper(int[]arr, int target, int low, int high){
        if(low > high) {
            return -1;
        }
        int mid = low + (high - low)/2;
        if (target > arr[mid]) {
            return helper(arr, target,mid+1, high);
        } else if(target < arr[mid]){
            return helper(arr, target, low, mid-1);
        }else {
            return mid;
        }
    }

    public static void main(String[] args) {
        int[]arr = {10,20, 30, 40, 50, 60};
        int target  = 50;
        BinarySearch binarySearch = new BinarySearch();
        int index = binarySearch.bSearchRecursive(arr, target);
        System.out.println("Index " + index);
    }
}
